use well_known[m.homeserver].base_url instead of home_server for cont…

teutat3s opened the pull request:
…acting the synapse server

Fixes #7 

Please excuse if the code isn't perfect, I'm still learning a lot and would be happy to improve it if possible, feedback is always welcome.

### Suggestions: 
- use `well_known[m.homeserver].base_url` from `matrix.YOUR.DOMAIN/_matrix/client/r0/login` instead of `home_server`
- this is to make synapse admin work with `matrix.YOUR.DOMAIN` instead of `YOUR.DOMAIN`
- also added a helper function to strip trailing slashes that can appear in `base_url`
- moved `https` checking into a function and use it when we first get the matrix URL, then store the correct `https://` URL in localstorage

rhaamo commented:
When accessing the login page without anything already in cache, `stripTrailingSlash` fails because `str` is null.
Didn't investigated much, I just added a return if `!str` to workaround.

teutat3s commented:
> When accessing the login page without anything already in cache, `stripTrailingSlash` fails because `str` is null.
> Didn't investigated much, I just added a return if `!str` to workaround.

Thank you for the hint, implemented as you suggested

Regarding this PR, I did improvements to my previous commit, now we check early if the URL was entered without `https` or if it's missing in the json we get from synapse (which probably shouldn't happen, but good to make sure) and add the correct url with `https` and without trailing spaces to localstorage, from where we can easily access it.

In my opinion using the well_known entry from the json for the homeserver address is the right way to go.

I'm looking forward for feedback.

awesome-manuel commented:
@teutat3s please add me or @awesome-michael as reviewer when you think it's ready.

teutat3s commented:
Thanks for the review, could you point me to the code where `home_server` is used - I don't know where that part of the logic happens.

teutat3s commented:
Testing create user form looks like it's not auto-detecting the `home_server`

<img width="288" alt="Screenshot 2020-02-12 at 14 38 33" src="https://user-images.githubusercontent.com/10206665/74339992-69e85d80-4da5-11ea-8e88-8b6030514596.png">

awesome-michael commented:
Right.. the parts that need the domain part of the user id are not (yet) in the master branch.

danjenkins commented:
Just come to this project and hit this exact issue - getting this merged in would be awesome

awesome-manuel commented on src/synapse/authProvider.js:
```diff
@@ -1,5 +1,22 @@
 import { fetchUtils } from "react-admin";
 
+const ensureHttpsForUrl = (url) => {
+  if (/^https:\/\//i.test(url)) {
```
This `if` is probably not required, since the `replace` will simply replace https:// by https://

teutat3s replied:
I thought of return early & often here - in a [regex tester](https://www.regextester.com/93652) `/http?:\/\//g` only matches `http://` but not `https://` so we only replace if necessary

awesome-manuel replied:
Right, so either remove the `?` after `http` in line 7 or put a `.` in front of it. Right now it makes the `p` optional which is not what we want.

awesome-manuel commented on src/synapse/authProvider.js:
```diff
@@ -13,16 +30,13 @@ const authProvider = {
-    }
+    const url = window.decodeURIComponent(homeserver);
+    const trimmed_url = url.trim().replace(/\s/g, "");
+    const homeserver_url = ensureHttpsForUrl(trimmed_url) + "/_matrix/client/r0/login";
```
Better rename to `login_api_url`

awesome-manuel commented on src/synapse/authProvider.js:
```diff
@@ -13,16 +30,13 @@ const authProvider = {
-    return fetchUtils.fetchJson(url, options).then(({ json }) => {
-      localStorage.setItem("home_server", json.home_server);
+    return fetchUtils.fetchJson(homeserver_url, options).then(({ json }) => {
+      const normalized_base_url = ensureHttpsForUrl(stripTrailingSlash(json.well_known["m.homeserver"].base_url))
```
If the `base_url` does not start with `https://` this is probably a server misconfiguration. So I don't think we should try to be smarter than the server admin here.

awesome-manuel commented on src/synapse/authProvider.js:
```diff
@@ -13,16 +30,13 @@ const authProvider = {
-      localStorage.setItem("home_server", json.home_server);
+    return fetchUtils.fetchJson(homeserver_url, options).then(({ json }) => {
+      const normalized_base_url = ensureHttpsForUrl(stripTrailingSlash(json.well_known["m.homeserver"].base_url))
+      localStorage.setItem("home_server", normalized_base_url);
```
Should be `localStorage.setItem("home_server_url", ...)`

teutat3s commented:
Thank you for all the feedback and sorry for the delay - I was away on holidays. 

Please check if everything is fine now - I tried implementing all requested changes, like adding `base_url` and smaller fixes that @awesome-manuel requested.
